### PR TITLE
Split Indices module from Balances

### LINKED ISSUE #919: Split indices out of balances module
Currently `AccountIndex` and `Address` are baked into the `balances` module. They should be split out into a lower-level `indices` module so that modules like `assets` can use account indices without needing to bring in `balances`.

### PR DESCRIPTION
**Pull request by gavofyork**
Closes #919 

Tasks
- [x] Split-out indices module
- [x] Split-out balances module
- [x] Move `Lookup` stuff into System module
- [x] Move traits out of Indices into System/Balances modules
- [x] Repot `IdentityLookup`
- [x] Fix up other modules & tests
- [x] Remove balances dependency from assets
- [x] Rejig node runtime
- [x] PR for substrate-node-template https://github.com/paritytech/substrate-node-template/pull/25
- [x] PR for polkadot https://github.com/paritytech/polkadot/pull/76
- [x] Tests in balances:
  - default_indexing_on_new_accounts_should_not_work
  - reserved_balance_should_prevent_reclaim_count

Also:
- [x] See if it's possible to move `Lookup` type param into system
- [x] Ensure all modules use `Lookup::Source` rather than `AccountId` in dispatchers.

### CODE REVIEW

File: srml/assets/src/lib.rs [hunk 1]
```diff
@@ -79,10 +81,13 @@ decl_module! {
 		}
 
 		/// Move some assets from one holder to another.
-		fn transfer(origin, id: AssetId, target: T::AccountId, amount: T::Balance) {
+		fn transfer(origin, id: Compact<AssetId>, target: <T::Lookup as StaticLookup>::Source, amount: <T::Balance as HasCompact>::Type) {
```
Comment (rphmeier) [inline srml/assets/src/lib.rs]: needs wrapping

File: srml/assets/src/lib.rs [hunk 2]
```diff
@@ -151,7 +157,7 @@ mod tests {
 	use substrate_primitives::{H256, Blake2Hasher};
 	// The testing primitives are very useful for avoiding having to work with signatures
 	// or public keys. `u64` is used as the `AccountId` and no `Signature`s are required.
-	use primitives::{BuildStorage, traits::{BlakeTwo256}, testing::{Digest, DigestItem, Header}};
+	use primitives::{BuildStorage, traits::{BlakeTwo256, IdentityLookup}, testing::{Digest, DigestItem, Header}};
```
Comment (rphmeier) [inline srml/assets/src/lib.rs]: needs wrapping

File: srml/system/src/lib.rs [hunk 3]
```diff
@@ -52,6 +52,28 @@ use runtime_io::{twox_128, TestExternalities, Blake2Hasher};
 #[cfg(any(feature = "std", test))]
 use substrate_primitives::ChangesTrieConfiguration;
 
+/// Handler for when a new account has been created.
+pub trait OnNewAccount<AccountId> {
```
Comment (rphmeier) [inline srml/system/src/lib.rs]: Does this and `IsDeadAccount` belong in `system` long-term? Is there a plan to move these to `indices` or some other `accounts-chain` thing?

**Comment by gavofyork** [comment]
@rphmeier ready for merge...

**Comment by gavofyork** [comment]
@jacogr you ready for this to go in?

**Comment by jacogr** [comment]
Yes, it can go. 

(We don't have a direct dependency, but rather an indirect one where we enhance with some derived functions, e.g. lookup AccountIndex from AccountId, etc. I guess that is similar to what it done in bonds. It needs some tweaking there, but that is the compat issue, nothing that stops this PR at all)

File: srml/system/src/lib.rs (hunk 3 above)
Comment (rphmeier) [reply srml/system/src/lib.rs]: Yes, this is what I mean -- I'm sure for @0x7CFE's UTXO prototype it doesn't help much to have AccountID in System for example, while other things in System are useful

**Pull request merged**